sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/repository/metric/InMemoryMetricsRepository.java

```diff
@@ -44,54 +43,71 @@
     /**
      * {@code app -> resource -> timestamp -> metric}
      */
-    private Map<String, Map<String, ConcurrentLinkedHashMap<Long, MetricEntity>>> allMetrics = new ConcurrentHashMap<>();
+    private Map<String, Map<String, LinkedHashMap<Long, MetricEntity>>> allMetrics = new ConcurrentHashMap<>();
 
+    private ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
```

Comment:
`readWriteLock` here will be escape?

Reply:
> `readWriteLock` here will be escape?

I'm sorry i don't understand what you mean. Adding final is unnecessary or there's any reason? Just an instance final variable which can't change.

Reply:
Haha, jason is a man who is very precise and detail oriented! If the `readWriteLock` add final decoration, then is the`allMetrics` variable also needed? Since they are both private, I think it's ok : )

Reply:
> > `readWriteLock` here will be escape?
> 
> I'm sorry i don't understand what you mean. Adding final is unnecessary or there's any reason? Just an instance final variable which can't change.

我的意思是你想通过final关键字确保`readWriteLock` 不会发对象逃逸吗。 `readWriteLock`  
是InMemoryMetricsRepository的私有属性并且没有暴露给外部访问的接口(getter/setter) 应该是不会发生逃逸的和也不会被外部所修改。所以加final关键字目的是？


I mean what you want to make sure is  'readWriteLock' prevent escape by using the final keyword?  ` readWriteLock `
Is InMemoryMetricsRepository private property and not exposed to the external access interface (getter/setter) . So  escape is not going to happen, and also not be modified by the external. So what is the purpose of adding the final keyword?

Reply:
> Haha, jason is a man who is very precise and detail oriented! If the `readWriteLock` add final decoration, then is the`allMetrics` variable also needed? Since they are both private, I think it's ok : )



> Haha, jason is a man who is very precise and detail oriented! If the `readWriteLock` add final decoration, then is the`allMetrics` variable also needed? Since they are both private, I think it's ok : )

I know， I just wondering if he had some extra reason.

Reply:
> > > `readWriteLock` here will be escape?
> > 
> > 
> > I'm sorry i don't understand what you mean. Adding final is unnecessary or there's any reason? Just an instance final variable which can't change.
> 
> 我的意思是你想通过final关键字确保`readWriteLock` 不会发对象逃逸吗。 `readWriteLock`
> 是InMemoryMetricsRepository的私有属性并且没有暴露给外部访问的接口(getter/setter) 应该是不会发生逃逸的和也不会被外部所修改。所以加final关键字目的是？
> 
> I mean what you want to make sure is 'readWriteLock' prevent escape by using the final keyword? `readWriteLock`
> Is InMemoryMetricsRepository private property and not exposed to the external access interface (getter/setter) . So escape is not going to happen, and also not be modified by the external. So what is the purpose of adding the final keyword?

Yep just to promise the purpose and you can take it as a kind of "documentation" like 

"I will not change anytime and every instance has a different one."

And maybe the compiler will replace it with fixed address but not variable sign when compiling.

Reply:
> Haha, jason is a man who is very precise and detail oriented! If the `readWriteLock` add final decoration, then is the`allMetrics` variable also needed? Since they are both private, I think it's ok : )

And `allMetrics` has no need to do the same thing in my opinion because it indeed can be replaced to another instance. But a lock cannot.

Reply:
> > > > `readWriteLock` here will be escape?
> > > 
> > > 
> > > I'm sorry i don't understand what you mean. Adding final is unnecessary or there's any reason? Just an instance final variable which can't change.
> > 
> > 
> > 我的意思是你想通过final关键字确保`readWriteLock` 不会发对象逃逸吗。 `readWriteLock`
> > 是InMemoryMetricsRepository的私有属性并且没有暴露给外部访问的接口(getter/setter) 应该是不会发生逃逸的和也不会被外部所修改。所以加final关键字目的是？
> > I mean what you want to make sure is 'readWriteLock' prevent escape by using the final keyword? `readWriteLock`
> > Is InMemoryMetricsRepository private property and not exposed to the external access interface (getter/setter) . So escape is not going to happen, and also not be modified by the external. So what is the purpose of adding the final keyword?
> 
> Yep just to promise the purpose and you can take it as a kind of "documentation" like
> 
> "I will not change anytime and every instance has a different one."
> 
> And maybe the compiler will replace it with fixed address but not variable sign when compiling.

ok, may be It is better to stress that a lock is final, when it is globally unique and won't be modified.

Reply:
> > > > > `readWriteLock` here will be escape?
> > > > 
> > > > 
> > > > I'm sorry i don't understand what you mean. Adding final is unnecessary or there's any reason? Just an instance final variable which can't change.
> > > 
> > > 
> > > 我的意思是你想通过final关键字确保`readWriteLock` 不会发对象逃逸吗。 `readWriteLock`
> > > 是InMemoryMetricsRepository的私有属性并且没有暴露给外部访问的接口(getter/setter) 应该是不会发生逃逸的和也不会被外部所修改。所以加final关键字目的是？
> > > I mean what you want to make sure is 'readWriteLock' prevent escape by using the final keyword? `readWriteLock`
> > > Is InMemoryMetricsRepository private property and not exposed to the external access interface (getter/setter) . So escape is not going to happen, and also not be modified by the external. So what is the purpose of adding the final keyword?
> > 
> > 
> > Yep just to promise the purpose and you can take it as a kind of "documentation" like
> > "I will not change anytime and every instance has a different one."
> > And maybe the compiler will replace it with fixed address but not variable sign when compiling.
> 
> ok, may be It is better to stress that a lock is final, when it is globally unique and won't be modified.

Yeah it's just a manner but not any constraint. Don't forget the TimeUtil issue and do you have any further question on it? It may effect the performance i think. Then we can get it done this time, right? @cdfive

Reply:
> Yep just to promise the purpose and you can take it as a kind of "documentation" like

"I will not change anytime and every instance has a different one."

And maybe the compiler will replace it with fixed address but not variable sign when compiling.

These two points are very precise, and I found the line in `CtSph` @jasonjoo2010 @linlinisme 
```java
private static final Object LOCK = new Object();
```

Reply:
> > Yep just to promise the purpose and you can take it as a kind of "documentation" like
> > "I will not change anytime and every instance has a different one."
> > And maybe the compiler will replace it with fixed address but not variable sign when compiling.
> 
> These two points are very precise, though not very sure which is better
> and I found the line in `CtSph` @jasonjoo2010 @linlinisme
> 
> ```java
> private static final Object LOCK = new Object();
> ```

Whether we should qualify it as `static` depends on actual design and `InMemoryMetricsRepository ` may need a per instance lock only. Though there may be only one instance in practice :)
And surely `static final` is the gramma to declare a "constant" so it's a good practice to name it all upper cases with underscore between each pair of word.
As per instance `final` both of styles is acceptable for me though.

Reply:
> > Yep just to promise the purpose and you can take it as a kind of "documentation" like
> > "I will not change anytime and every instance has a different one."
> > And maybe the compiler will replace it with fixed address but not variable sign when compiling.
> 
> These two points are very precise, though not very sure which is better
> and I found the line in `CtSph` @jasonjoo2010 @linlinisme
> 
> ```java
> private static final Object LOCK = new Object();
> ```

Whether we should qualify it as `static` depends on actual design and `InMemoryMetricsRepository ` may need a per instance lock only. Though there may be only one instance in practice :)
And surely `static final` is the gramma to declare a "constant" so name it all upper cases with 

> > Yep just to promise the purpose and you can take it as a kind of "documentation" like
> > "I will not change anytime and every instance has a different one."
> > And maybe the compiler will replace it with fixed address but not variable sign when compiling.
> 
> These two points are very precise, though not very sure which is better
> and I found the line in `CtSph` @jasonjoo2010 @linlinisme
> 
> ```java
> private static final Object LOCK = new Object();
> ```

By the way we should speed up your design in another PR & issue because it's quite important maybe. I have make several proposals including the one based on yours :)

Reply:
感谢Jason耐心详实的说明。